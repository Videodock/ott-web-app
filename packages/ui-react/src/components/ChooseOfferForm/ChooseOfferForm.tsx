import React from 'react';
import { useTranslation } from 'react-i18next';
import classNames from 'classnames';
import type { FormErrors } from '@jwp/ott-common/types/form';
import type { Offer, ChooseOfferFormData } from '@jwp/ott-common/types/checkout';
import { useConfigStore } from '@jwp/ott-common/src/stores/ConfigStore';
import { getOfferPrice, isSVODOffer } from '@jwp/ott-common/src/utils/offers';
import { testId } from '@jwp/ott-common/src/utils/common';
import CheckCircle from '@jwp/ott-theme/assets/icons/check_circle.svg?react';

import Button from '../Button/Button';
import FormFeedback from '../FormFeedback/FormFeedback';
import DialogBackButton from '../DialogBackButton/DialogBackButton';
import LoadingOverlay from '../LoadingOverlay/LoadingOverlay';
import Icon from '../Icon/Icon';

import styles from './ChooseOfferForm.module.scss';

type OfferBoxProps = {
  offer: Offer;
  selected: boolean;
} & Pick<Props, 'onChange'>;

const OfferBox: React.FC<OfferBoxProps> = ({ offer, selected, onChange }: OfferBoxProps) => {
  const { t } = useTranslation('account');

  const getFreeTrialText = (offer: Offer) => {
    if (offer.freeDays) {
      return t('choose_offer.benefits.first_days_free', { count: offer.freeDays });
    } else if (offer.freePeriods) {
      // t('periods.day', { count })
      // t('periods.week', { count })
      // t('periods.month', { count })
      // t('periods.year', { count })
      const period = t(`periods.${offer.period}`, { count: offer.freePeriods });

      return t('choose_offer.benefits.first_periods_free', { count: offer.freePeriods, period });
    }

    return null;
  };

  const renderOption = ({ title, periodString, secondBenefit }: { title: string; periodString?: string; secondBenefit?: string }) => (
    <div className={styles.offer} role="option">
      <input
        className={styles.radio}
        onChange={onChange}
        type="radio"
        name={'selectedOfferId'}
        value={offer.offerId}
        id={offer.offerId}
        checked={selected}
        data-testid={testId(title)}
      />
      <label className={styles.label} htmlFor={offer.offerId}>
        <h2 className={styles.offerTitle}>{title}</h2>
        <hr className={styles.offerDivider} />
        <ul className={styles.offerBenefits}>
          {offer.freeDays || offer.freePeriods ? (
            <li>
              <Icon icon={CheckCircle} /> {getFreeTrialText(offer)}
            </li>
          ) : null}

          {!!secondBenefit && (
            <li>
              <Icon icon={CheckCircle} /> {secondBenefit}
            </li>
          )}
          <li>
            <Icon icon={CheckCircle} /> {t('choose_offer.benefits.watch_on_all_devices')}
          </li>
        </ul>
        <div className={styles.fill} />
        <div className={styles.offerPrice}>
          {getOfferPrice(offer)} {!!periodString && <small>/{periodString}</small>}
        </div>
      </label>
    </div>
  );

  if (isSVODOffer(offer)) {
    const isMonthly = offer.period === 'month';

    return renderOption({
      title: isMonthly ? t('choose_offer.monthly') : t('choose_offer.yearly'),
      secondBenefit: t('choose_offer.benefits.cancel_anytime'),
      periodString: isMonthly ? t('periods.month') : t('periods.year'),
    });
  }

  return renderOption({
    title: offer.offerTitle,
    secondBenefit:
      !!offer.durationPeriod && !!offer.durationAmount
        ? t('choose_offer.tvod_access', { period: offer.durationPeriod, count: offer.durationAmount })
        : undefined,
  });
};

type Props = {
  values: ChooseOfferFormData;
  errors: FormErrors<ChooseOfferFormData>;
  onChange: React.ChangeEventHandler<HTMLInputElement | HTMLTextAreaElement>;
  onSubmit: React.FormEventHandler<HTMLFormElement>;
  onBackButtonClickHandler?: () => void;
  offers: Offer[];
  showOfferTypeSwitch: boolean;
  submitting: boolean;
};

const ChooseOfferForm: React.FC<Props> = ({ values, errors, submitting, offers, showOfferTypeSwitch, onChange, onSubmit, onBackButtonClickHandler }: Props) => {
  const siteName = useConfigStore((s) => s.config.siteName);
  const { t } = useTranslation('account');
  const { selectedOfferType, selectedOfferId } = values;

  return (
    <form onSubmit={onSubmit} data-testid={testId('choose-offer-form')} noValidate>
      {onBackButtonClickHandler ? <DialogBackButton onClick={onBackButtonClickHandler} /> : null}
      <h1 className={styles.title}>{t('choose_offer.title')}</h1>
      <p className={styles.subtitle}>{t('choose_offer.watch_this_on_platform', { siteName })}</p>
      {errors.form ? <FormFeedback variant="error">{errors.form}</FormFeedback> : null}
      {showOfferTypeSwitch && (
        <div className={styles.offerGroupSwitch}>
          <input
            className={styles.radio}
            onChange={onChange}
            type="radio"
            name="selectedOfferType"
            id="svod"
            value="svod"
            checked={selectedOfferType === 'svod'}
          />
          <label className={classNames(styles.label, styles.offerGroupLabel)} htmlFor="svod">
            {t('choose_offer.subscription')}
          </label>
          <input
            className={styles.radio}
            onChange={onChange}
            type="radio"
            name="selectedOfferType"
            id="tvod"
            value="tvod"
            checked={selectedOfferType === 'tvod'}
          />
          <label className={classNames(styles.label, styles.offerGroupLabel)} htmlFor="tvod">
            {t('choose_offer.one_time_only')}
          </label>
        </div>
      )}
      <div className={styles.offers}>
        {!offers.length ? (
          <p>{t('choose_offer.no_pricing_available')}</p>
        ) : (
          offers.map((offer) => <OfferBox key={offer.offerId} offer={offer} selected={selectedOfferId === offer.offerId} onChange={onChange} />)
        )}
      </div>
      {submitting && <LoadingOverlay transparentBackground inline />}
      <Button label={t('choose_offer.continue')} disabled={submitting || !offers.length} variant="contained" color="primary" type="submit" fullWidth />
    </form>
  );
};
export default ChooseOfferForm;
