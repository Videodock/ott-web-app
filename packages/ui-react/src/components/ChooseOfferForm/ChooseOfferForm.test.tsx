import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import type { Offer } from '@jwp/ott-common/types/checkout';
import monthlyOffer from '@jwp/ott-testing/fixtures/monthlyOffer.json';
import yearlyOffer from '@jwp/ott-testing/fixtures/yearlyOffer.json';
import tvodOffer from '@jwp/ott-testing/fixtures/tvodOffer.json';

import ChooseOfferForm from './ChooseOfferForm';

const svodOffers = [monthlyOffer, yearlyOffer] as Offer[];
const tvodOffers = [tvodOffer] as Offer[];

describe('<OffersForm>', () => {
  test('renders and matches snapshot', () => {
    const { container } = render(
      <ChooseOfferForm
        values={{ offerId: 'S916977979_NL' }}
        errors={{}}
        onChange={vi.fn()}
        onSubmit={vi.fn()}
        submitting={false}
        offers={svodOffers}
        offerType={'svod'}
        setOfferType={vi.fn()}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  test('renders and matches snapshot', () => {
    const { container } = render(
      <ChooseOfferForm
        values={{ offerId: 'R892134629_NL' }}
        errors={{}}
        onChange={vi.fn()}
        onSubmit={vi.fn()}
        submitting={false}
        offers={tvodOffers}
        offerType={'tvod'}
        setOfferType={vi.fn()}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  test('checks the monthly offer correctly', () => {
    const { getByTestId } = render(
      <ChooseOfferForm
        values={{ offerId: 'S916977979_NL' }}
        errors={{}}
        onChange={vi.fn()}
        onSubmit={vi.fn()}
        submitting={false}
        offers={svodOffers}
        offerType={'svod'}
        setOfferType={vi.fn()}
      />,
    );

    expect(getByTestId('choose_offer.monthly')).toBeChecked();
  });

  test('checks the yearly offer correctly', () => {
    const { getByTestId } = render(
      <ChooseOfferForm
        values={{ offerId: 'S345569153_NL' }}
        errors={{}}
        onChange={vi.fn()}
        onSubmit={vi.fn()}
        submitting={false}
        offers={svodOffers}
        offerType={'svod'}
        setOfferType={vi.fn()}
      />,
    );

    expect(getByTestId('choose_offer.yearly')).toBeChecked();
  });

  test('checks the tvod offer correctly', () => {
    const { getByTestId } = render(
      <ChooseOfferForm
        values={{ offerId: 'R892134629_NL' }}
        errors={{}}
        onChange={vi.fn()}
        onSubmit={vi.fn()}
        submitting={false}
        offers={tvodOffers}
        offerType={'tvod'}
        setOfferType={vi.fn()}
      />,
    );

    expect(getByTestId('One Time - TVOD offer')).toBeChecked();
  });

  test('calls the onChange callback when changing the offer', () => {
    const onChange = vi.fn();
    const { getByTestId } = render(
      <ChooseOfferForm
        values={{ offerId: 'S916977979_NL' }}
        errors={{}}
        onChange={onChange}
        onSubmit={vi.fn()}
        submitting={false}
        offers={svodOffers}
        offerType={'svod'}
        setOfferType={vi.fn()}
      />,
    );

    fireEvent.click(getByTestId('choose_offer.yearly'));

    expect(onChange).toBeCalled();
  });

  test('calls the onSubmit callback when submitting the form', () => {
    const onSubmit = vi.fn();
    const { getByTestId } = render(
      <ChooseOfferForm
        values={{ offerId: 'S916977979_NL' }}
        errors={{}}
        onChange={vi.fn()}
        onSubmit={onSubmit}
        submitting={false}
        offers={svodOffers}
        offerType={'svod'}
        setOfferType={vi.fn()}
      />,
    );

    fireEvent.submit(getByTestId('choose-offer-form'));

    expect(onSubmit).toBeCalled();
  });
});
