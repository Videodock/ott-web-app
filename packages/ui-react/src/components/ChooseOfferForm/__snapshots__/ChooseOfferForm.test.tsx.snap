// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<OffersForm> > renders and matches snapshot 1`] = `
<div>
  <form
    data-testid="choose-offer-form"
    novalidate=""
  >
    <h1
      class="_title_f7961f"
    >
      choose_offer.title
    </h1>
    <p
      class="_subtitle_f7961f"
    >
      choose_offer.watch_this_on_platform
    </p>
    <div
      class="_offerGroupSwitch_f7961f"
    >
      <input
        checked=""
        class="_radio_f7961f"
        id="svod"
        name="selectedOfferType"
        type="radio"
        value="svod"
      />
      <label
        class="_label_f7961f _offerGroupLabel_f7961f"
        for="svod"
      >
        choose_offer.subscription
      </label>
      <input
        class="_radio_f7961f"
        id="tvod"
        name="selectedOfferType"
        type="radio"
        value="tvod"
      />
      <label
        class="_label_f7961f _offerGroupLabel_f7961f"
        for="tvod"
      >
        choose_offer.one_time_only
      </label>
    </div>
    <div
      class="_offers_f7961f"
    >
      <div
        aria-labelledby="title-S916977979_NL"
        class="_offer_f7961f"
      >
        <input
          checked=""
          class="_radio_f7961f"
          data-testid="S916977979_NL"
          id="S916977979_NL"
          name="selectedOfferId"
          type="radio"
          value="S916977979_NL"
        />
        <div
          class="_label_f7961f"
        >
          <label
            for="S916977979_NL"
          >
            <h2
              class="_offerTitle_f7961f"
              id="title-S916977979_NL"
            >
              choose_offer.monthly
            </h2>
            <span
              class="hidden"
            >
              .
            </span>
            <hr
              aria-hidden="true"
              class="_offerDivider_f7961f"
            />
            <ul
              class="_offerBenefits_f7961f"
            >
              <li>
                <svg
                  aria-hidden="true"
                  class="_icon_585b29"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 2C6.5 2 2 6.5 2 12S6.5 22 12 22 22 17.5 22 12 17.5 2 12 2M10 17L5 12L6.41 10.59L10 14.17L17.59 6.58L19 8L10 17Z"
                  />
                </svg>
                choose_offer.benefits.first_periods_free
                <span
                  class="hidden"
                >
                  .
                </span>
              </li>
              <li>
                <svg
                  aria-hidden="true"
                  class="_icon_585b29"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 2C6.5 2 2 6.5 2 12S6.5 22 12 22 22 17.5 22 12 17.5 2 12 2M10 17L5 12L6.41 10.59L10 14.17L17.59 6.58L19 8L10 17Z"
                  />
                </svg>
                choose_offer.benefits.cancel_anytime
                <span
                  class="hidden"
                >
                  .
                </span>
              </li>
              <li>
                <svg
                  aria-hidden="true"
                  class="_icon_585b29"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 2C6.5 2 2 6.5 2 12S6.5 22 12 22 22 17.5 22 12 17.5 2 12 2M10 17L5 12L6.41 10.59L10 14.17L17.59 6.58L19 8L10 17Z"
                  />
                </svg>
                choose_offer.benefits.watch_on_all_devices
                <span
                  class="hidden"
                >
                  .
                </span>
              </li>
            </ul>
            <div
              class="_fill_f7961f"
            />
            <div
              class="_offerPrice_f7961f"
            >
              € 6,99
               
              <small>
                /
                periods.month
              </small>
            </div>
          </label>
        </div>
      </div>
      <div
        aria-labelledby="title-S345569153_NL"
        class="_offer_f7961f"
      >
        <input
          class="_radio_f7961f"
          data-testid="S345569153_NL"
          id="S345569153_NL"
          name="selectedOfferId"
          type="radio"
          value="S345569153_NL"
        />
        <div
          class="_label_f7961f"
        >
          <label
            for="S345569153_NL"
          >
            <h2
              class="_offerTitle_f7961f"
              id="title-S345569153_NL"
            >
              choose_offer.yearly
            </h2>
            <span
              class="hidden"
            >
              .
            </span>
            <hr
              aria-hidden="true"
              class="_offerDivider_f7961f"
            />
            <ul
              class="_offerBenefits_f7961f"
            >
              <li>
                <svg
                  aria-hidden="true"
                  class="_icon_585b29"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 2C6.5 2 2 6.5 2 12S6.5 22 12 22 22 17.5 22 12 17.5 2 12 2M10 17L5 12L6.41 10.59L10 14.17L17.59 6.58L19 8L10 17Z"
                  />
                </svg>
                choose_offer.benefits.first_days_free
                <span
                  class="hidden"
                >
                  .
                </span>
              </li>
              <li>
                <svg
                  aria-hidden="true"
                  class="_icon_585b29"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 2C6.5 2 2 6.5 2 12S6.5 22 12 22 22 17.5 22 12 17.5 2 12 2M10 17L5 12L6.41 10.59L10 14.17L17.59 6.58L19 8L10 17Z"
                  />
                </svg>
                choose_offer.benefits.cancel_anytime
                <span
                  class="hidden"
                >
                  .
                </span>
              </li>
              <li>
                <svg
                  aria-hidden="true"
                  class="_icon_585b29"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 2C6.5 2 2 6.5 2 12S6.5 22 12 22 22 17.5 22 12 17.5 2 12 2M10 17L5 12L6.41 10.59L10 14.17L17.59 6.58L19 8L10 17Z"
                  />
                </svg>
                choose_offer.benefits.watch_on_all_devices
                <span
                  class="hidden"
                >
                  .
                </span>
              </li>
            </ul>
            <div
              class="_fill_f7961f"
            />
            <div
              class="_offerPrice_f7961f"
            >
              € 50,00
               
              <small>
                /
                periods.year
              </small>
            </div>
          </label>
        </div>
      </div>
    </div>
    <button
      aria-disabled="false"
      class="_button_f8f296 _primary_f8f296 _fullWidth_f8f296"
      type="submit"
    >
      <span>
        choose_offer.continue
      </span>
    </button>
  </form>
</div>
`;

exports[`<OffersForm> > renders and matches snapshot 2`] = `
<div>
  <form
    data-testid="choose-offer-form"
    novalidate=""
  >
    <h1
      class="_title_f7961f"
    >
      choose_offer.title
    </h1>
    <p
      class="_subtitle_f7961f"
    >
      choose_offer.watch_this_on_platform
    </p>
    <div
      class="_offerGroupSwitch_f7961f"
    >
      <input
        class="_radio_f7961f"
        id="svod"
        name="selectedOfferType"
        type="radio"
        value="svod"
      />
      <label
        class="_label_f7961f _offerGroupLabel_f7961f"
        for="svod"
      >
        choose_offer.subscription
      </label>
      <input
        checked=""
        class="_radio_f7961f"
        id="tvod"
        name="selectedOfferType"
        type="radio"
        value="tvod"
      />
      <label
        class="_label_f7961f _offerGroupLabel_f7961f"
        for="tvod"
      >
        choose_offer.one_time_only
      </label>
    </div>
    <div
      class="_offers_f7961f"
    >
      <div
        aria-labelledby="title-R892134629_NL"
        class="_offer_f7961f"
      >
        <input
          checked=""
          class="_radio_f7961f"
          data-testid="R892134629_NL"
          id="R892134629_NL"
          name="selectedOfferId"
          type="radio"
          value="R892134629_NL"
        />
        <div
          class="_label_f7961f"
        >
          <label
            for="R892134629_NL"
          >
            <h2
              class="_offerTitle_f7961f"
              id="title-R892134629_NL"
            >
              One Time - TVOD offer
            </h2>
            <span
              class="hidden"
            >
              .
            </span>
            <hr
              aria-hidden="true"
              class="_offerDivider_f7961f"
            />
            <ul
              class="_offerBenefits_f7961f"
            >
              <li>
                <svg
                  aria-hidden="true"
                  class="_icon_585b29"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 2C6.5 2 2 6.5 2 12S6.5 22 12 22 22 17.5 22 12 17.5 2 12 2M10 17L5 12L6.41 10.59L10 14.17L17.59 6.58L19 8L10 17Z"
                  />
                </svg>
                choose_offer.benefits.watch_on_all_devices
                <span
                  class="hidden"
                >
                  .
                </span>
              </li>
            </ul>
            <div
              class="_fill_f7961f"
            />
            <div
              class="_offerPrice_f7961f"
            >
              € 2,90
               
            </div>
          </label>
        </div>
      </div>
    </div>
    <button
      aria-disabled="false"
      class="_button_f8f296 _primary_f8f296 _fullWidth_f8f296"
      type="submit"
    >
      <span>
        choose_offer.continue
      </span>
    </button>
  </form>
</div>
`;
